### ISSUE TITLE
Replace hard-coded RelationshipType enum with something more flexible e.g. Term

### ISSUE THREAD

**Issue by julesjacobsen** [monarch-initiative/phenol#184]
Currently phenol hard codes relation types with the intent of doing some minimum Q/C, so that all of the relation types are at least in principle computable (the alternative would be to create a relation on the fly or to map all relations to some generic relation). The disadvantage is that phenol crashes if it gets a relation it has never seen. For now, we can just add this relation, but perhaps there is a better strategy moving forward.

_Originally posted by @pnrobinson in https://github.com/monarch-initiative/phenol/issues/163#issuecomment-446003855_

**Comment by julesjacobsen** [comment]
Currently the ```OboGraphDocumentAdaptor``` takes a limited configuration in the form of a CurieUtil and a set of term id prefixes in order to filter out unwanted/unknown nodes. For the HPO and MPO at least there is no additional configuration required in order to get the graph required for Phenol type operations. 

```java
// The HPO is in the default  curie map and only contains known relationships / HP terms
Ontology hpoOntology = OntologyLoader.loadOntology(hpoFile)
```

For the GO however it becomes more complicated with the addition or RO and BFO terms with relationships outside of the current enum.

```java
// The GO is in the default curie map but also contains BFO and RO terms with unknown relationships
// we want to ignore these so here we specify the term prefixes we want to use. It has three possible root
// nodes (biological_process, cellular_component, biological_function) so an artificial GO:0000000 root is
// added
Ontology goOntology = OntologyLoader.loadOntology(goFile, "GO")
```

in response to issue #163 the ```RelationshipType``` would no longer explode if given an unknown relationship and would instead return a ```RelationshipType .UNKNOWN```. The issue now is that people could mistakenly try to open an ontology and then have a bunch of meaningless relationships. So one question is should these be removed by default? (https://github.com/monarch-initiative/phenol/issues/163#issuecomment-446553007 and https://github.com/monarch-initiative/phenol/issues/163#issuecomment-447076836 suggest yes)

If removal default removal of ```RelationshipType .UNKNOWN``` is the generally accepted way to go then ought there be cases where we want to be able to define all the required nodes and relationship types.

How far do we want to go with this? Could it be the case where we might want to work with a full and complicated graph exactly as it comes out of ontobio?

ECTO being one example of containing too much information. Loading it like this gives us a sinle graph of is_a relations an only ECTO nodes: 

```java
// ECTO isn't mapped in the default Curie mappings, so we need to add it here (the PURL isn't correct)
CurieUtil curieUtil = CurieUtilBuilder.withDefaultsAnd(ImmutableMap.of("ECTO", "http://purl.obolibrary.org/obo/ECTO_"));

// ECTO also contains a bunch of unknown relationships so we're going to simplify this graph by only
// loading ECTO nodes (this ignores the true root term XCO:0000000) and other nodes from CHEBI,
// BFO and UBERON among others.
Ontology ecto = OntologyLoader.loadOntology(ectoFile, curieUtil, "ECTO");
```
However loading this, produces a lot of CHEBI, UBERON, BFO, RO nodes and a lot of UNKNOWN relationship types:
```java
// ECTO isn't mapped in the default Curie mappings, so we need to add it here (the PURL isn't correct)
CurieUtil curieUtil = CurieUtilBuilder.withDefaultsAnd(ImmutableMap.of("ECTO", "http://purl.obolibrary.org/obo/ECTO_"));

// This now contains most of what was in the original graph loaded from ontolib
Ontology ecto = OntologyLoader.loadOntology(ectoFile, curieUtil);
```

Bearing in mind [Chris' comments](https://github.com/monarch-initiative/phenol/issues/163#issuecomment-447315278), I think it should now be relatively easy to provide the ```OboGraphDocumentAdaptor``` with an ```OntologyConfiguration``` (working name) which is an interface for enabling users to specify exactly what they want to see in the output. This will require users to be aware that the ontology they're loading might just enable them to create a horrible mess if they wish, but ought to stick to a safe default of only ```Node.RDFTYPES.CLASS```, and is_a relationships for example. During this work the RelationshipType ought to be migrated to something like a Term/TermId to enable any random ontology to be loaded without any preconceptions, or easily configured to filter out unwanted parts.

So using the above examples, HPO would require no change, GO might have a preset if we want to consider that a 'supported' ontology.

```java
Ontology hpoOntology = OntologyLoader.loadOntology(hpoFile)
// would be this under the hood:
Ontology hpoOntology = OntologyLoader.loadOntology(hpoFile, OntologyConfiguration.default());

// special GO config to replicate current setup returns a GoOntologyConfiguration 
Ontology goSubsetOntology = OntologyLoader.loadOntology(goFile, OntologyConfiguration.forGo())

// Empty configuration to reproduce the original original graph loaded from ontobio
Ontology goFullOntology = OntologyLoader.loadOntology(goFile, OntologyConfiguration.empty())

// ECTO isn't mapped in the default Curie mappings, so we need to add it here (the PURL isn't correct)
CurieUtil curieUtil = CurieUtilBuilder.withDefaultsAnd(ImmutableMap.of("ECTO", "http://purl.obolibrary.org/obo/ECTO_"));
OntologyConfiguration userDefinedConfiguration = OntologyConfiguration .builder()...build();

// This now contains most of what was in the original graph loaded from ontobio
Ontology ecto = OntologyLoader.loadOntology(ectoFile, userDefinedConfiguration);
```

**Comment by pnrobinson** [comment]
+1
It would be great for us to define "sensible" defaults for ontologies we plan to use ourselves, and to document the library sufficiently that others will understand how to deviate from the default. 

OntologyConfiguration.forGo()
OntologyConfiguration.forEcto()
.... for parts of the NCIT too?

would be great!

**Comment by pnrobinson** [comment]
@julesjacobsen Should we move forward with this? It doesn't look like you checked this in or am I missing something?

**Comment by julesjacobsen** [comment]
Um, no. Sorry - got involved in other things. I can start working on this next week.

**Comment by pnrobinson** [comment]
@julesjacobsen should we move ahead on this? We can split the coding if that would be a good idea, but perhaps we should touch bases about the best code architecture. I would like to "release" phenol on the HPO website as one of the tools, and perhaps bump the version to 1.4, after we have addressed the current issues, and this seems to be the major issue still outstanding.

**Comment by julesjacobsen** [comment]
@pnrobinson I'd say there are actually two related issues here, which is my fault as I went off on a tangent with the proposed ```OntologyConfiguration```. I think this is a relatively simple change, but one which will dramatically improve the usability and reduce surprises. I'll get working on this today.

The second, more fundamental issue is the replacement of the hard-coded relationship types, which is what this ticket is really about. As I already mentioned, I have a working change already, but it might be advantageous to switch to modelling things the same way as ontobio.

**Comment by pnrobinson** [comment]
@julesjacobsen This is perhaps related to #223 do we need to work on the underlying modeling. The class logic of phenol goes back to very old decisions and some of it is orthogonal to any current use case (mainly HPO). Can we discuss this?

**Comment by julesjacobsen** [comment]
@pnrobinson can you close this issue if you judge it sufficiently addressed by e3ff273. I've opened a new issue to address the `OntologyConfiguration` comments.

**Comment by pnrobinson** [comment]
Looks good to go!
